packages/twenty-server/src/engine/core-modules/auth/services/sign-in-up.service.ts

```diff
@@ -214,18 +213,39 @@ export class SignInUpService {
     return await this.userRepository.save(userToCreate);
   }
 
+  private async isWorkspaceReadyForSignInUpOrThrow(
+    workspace: Workspace,
+    user: ExistingUserOrPartialUserWithPicture,
+  ) {
+    if (workspace.activationStatus === WorkspaceActivationStatus.ACTIVE) return;
+
+    if (user.userData.type !== 'existingUser') {
+      throw new AuthException(
+        'Workspace is not ready to welcome new members',
+        AuthExceptionCode.FORBIDDEN_EXCEPTION,
+      );
+    }
+
+    const userWorkspaceExists =
+      await this.userWorkspaceService.checkUserWorkspaceExists(
+        user.userData.existingUser.id,
+        workspace.id,
+      );
+
+    if (!userWorkspaceExists) {
+      throw new AuthException(
+        'Workspace is not ready to welcome new members',
```

Comment:
we should throw a more explicit message: 'User is not part of the workspace' in this case

Reply:
If a user uses an invitation or a public link or whatever invite mode, he will receive an inaccurate message because he is trying to join the workspace. Hence, he knows that the "User is not part of the workspace" but he can't join the workspace because the "Workspace is not ready to welcome new members".

Reply:
Yes, I thought that this will be captured by the user.userData.type !== 'existingUser' case but indeed, we could have an existing user joining another workspace. In this case we should distinguish on the intent (invitation or not) rather than on database state